<a href="https://colab.research.google.com/github/venkatamogili/Analyzing_CustomerOrder_Using-_Python/blob/main/Analyzing_Customer_Orders_Using_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Situation:**
You are working as a data analyst for an e-commerce company. Your task is to process
and analyze customer orders to generate meaningful insights. Your company sells a
variety of products across different categories, such as Electronics, Clothing, and Home
Essentials. You need to determine which products are frequently purchased, classify
customers based on their total spending, and analyze the most profitable product
categories.
To achieve this, you will store, categorize, and analyze customer purchase data using
Python's built-in data structures (lists, tuples, dictionaries, and sets) and control
structures (loops and conditionals). The final output will help business managers make
data-driven decisions regarding marketing strategies and inventory management.

**Task 1:Store customer orders**

*    Create a list of customer names
*   Store each customer's order details (customer name, product, price, category) as
tuples inside a list
*   Use a dictionary where keys are customer names and values are lists of ordered
products



In [ ]:

# 1) Create a list of customer names
customers = [
    "Alex Johnson",
    "Priya Singh",
    "David Chen",
    "Maria Gomez",
    "Sarah Lee",
    "Michael Brown",
]

# 2) Store each customer's order details as tuples inside a list
#    Tuple format: (customer_name, product, price, category)
orders = [
    ("Alex Johnson",  "USB-C Cable",           9.99,  "Electronics"),
    ("Priya Singh",   "Hoodie",               39.00,  "Clothing"),
    ("David Chen",    "Coffee Maker",         59.00,  "Home Essentials"),
    ("Maria Gomez",   "Sneakers",             99.00,  "Clothing"),
    ("Sarah Lee",     "Pillow Set (2)",       17.99,  "Home Essentials"),
    ("Michael Brown", "Bluetooth Headphones", 69.00,  "Electronics"),
    ("Michael Brown", "USB-C Cable",           8.49,  "Electronics"),
    ("Alex Johnson",  "Pillow Set (2)",       16.99,  "Home Essentials"),
]

# 3) Use a dictionary: keys = customer names, values = lists of ordered products
#    We'll build this dictionary from the orders list.
customer_to_products = {}

for (cust, product, price, category) in orders:
    # Initialize the list for a customer if not present
    if cust not in customer_to_products:
        customer_to_products[cust] = []
    # Append only the product name (per your requirement)
    customer_to_products[cust].append(product)

# ---------- Optional helpers & examples ----------

def add_customer(customers_list, name):
    """Add a new customer if not already in the list."""
    if name not in customers_list:
        customers_list.append(name)

def add_order(orders_list, cust, product, price, category):
    """Add a new order as a tuple and return it."""
    order_tuple = (cust, product, price, category)
    orders_list.append(order_tuple)
    return order_tuple

def rebuild_customer_to_products(orders_list):
    """Rebuild the customer→products dictionary from the orders list."""
    mapping = {}
    for (cust, product, price, category) in orders_list:
        if cust not in mapping:
            mapping[cust] = []
        mapping[cust].append(product)
    return mapping



**Task 2 : Classify products by category**

• Use a dictionary to map each product to its respective category

• Create a set of unique product categories

• Display all available product categories

In [ ]:

# Dictionary mapping product → category
product_category = {
    "USB-C Cable": "Electronics",
    "Bluetooth Headphones": "Electronics",
    "Smartphone": "Electronics",

    "Hoodie": "Clothing",
    "Jeans": "Clothing",
    "Sneakers": "Clothing",

    "Coffee Maker": "Home Essentials",
    "Pillow Set": "Home Essentials",
    "Air Fryer": "Home Essentials",
}


In [ ]:

# Create a set of unique categories
unique_categories = set(product_category.values())


In [ ]:

print("Available Product Categories:")
for category in unique_categories:
    print("-", category)


Available Product Categories:
- Electronics
- Clothing
- Home Essentials


**3. Analyze customer orders**

• Use a loop to calculate the total amount each customer spends

• If the total purchase value is above $100, classify the customer as a high-value buyer

• If it is between $50  and  $100, classify the customer as a moderate buyer

• If it is below $50, classify them as a low-value buyer


In [ ]:

# Sample orders as tuples: (customer_name, product, price, category)
orders = [
    ("Alex Johnson",  "USB-C Cable",           9.99,  "Electronics"),
    ("Priya Singh",   "Hoodie",               39.00,  "Clothing"),
    ("David Chen",    "Coffee Maker",         59.00,  "Home Essentials"),
    ("Maria Gomez",   "Sneakers",             99.00,  "Clothing"),
    ("Sarah Lee",     "Pillow Set (2)",       17.99,  "Home Essentials"),
    ("Michael Brown", "Bluetooth Headphones", 69.00,  "Electronics"),
    ("Michael Brown", "USB-C Cable",           8.49,  "Electronics"),
    ("Alex Johnson",  "Pillow Set (2)",       16.99,  "Home Essentials"),
    ("Alex Johnson",  "USB-C Cable",           8.49,  "Electronics"),
    ("Sarah Lee",     "USB-C Cable",           7.99,  "Electronics"),
]

# 1) Loop to calculate total amount each customer spends
customer_totals = {}  # dict: customer_name -> total_spend

for (cust, product, price, category) in orders:
    if cust not in customer_totals:
        customer_totals[cust] = 0.0
    customer_totals[cust] += price

# 2) Classify customers by total spend using given thresholds
#    High-value: > $100
#    Moderate:   $50–$100 (inclusive)
#    Low-value:  < $50
customer_segments = {}  # dict: customer_name -> segment_label

for cust, total in customer_totals.items():
    if total > 100.0:
        segment = "High-value buyer"
    elif 50.0 <= total <= 100.0:
        segment = "Moderate buyer"
    else:  # total < 50.0
        segment = "Low-value buyer"
    customer_segments[cust] = segment

# 3) Display results
print("Customer Total Spend & Segment")
print("--------------------------------")
for cust in sorted(customer_totals.keys()):
    total = customer_totals[cust]
    segment = customer_segments[cust]
    print(f"{cust:15s}  Total: ${total:6.2f}  →  {segment}")


Customer Total Spend & Segment
--------------------------------
Alex Johnson     Total: $ 35.47  →  Low-value buyer
David Chen       Total: $ 59.00  →  Moderate buyer
Maria Gomez      Total: $ 99.00  →  Moderate buyer
Michael Brown    Total: $ 77.49  →  Moderate buyer
Priya Singh      Total: $ 39.00  →  Low-value buyer
Sarah Lee        Total: $ 25.98  →  Low-value buyer


**4. Generate business insights**

• Calculate the total revenue per product category and store it in a dictionary

• Extract unique products from all orders using a set

• Use a list comprehension to find all customers who purchased electronics

• Identify the top three highest-spending customers using sorting


In [ ]:

# Sample orders as tuples: (customer_name, product, price, category)
orders = [
    ("Alex Johnson",  "USB-C Cable",           9.99,  "Electronics"),
    ("Priya Singh",   "Hoodie",               39.00,  "Clothing"),
    ("David Chen",    "Coffee Maker",         59.00,  "Home Essentials"),
    ("Maria Gomez",   "Sneakers",             99.00,  "Clothing"),
    ("Sarah Lee",     "Pillow Set (2)",       17.99,  "Home Essentials"),
    ("Michael Brown", "Bluetooth Headphones", 69.00,  "Electronics"),
    ("Michael Brown", "USB-C Cable",           8.49,  "Electronics"),
    ("Alex Johnson",  "Pillow Set (2)",       16.99,  "Home Essentials"),
    ("Alex Johnson",  "USB-C Cable",           8.49,  "Electronics"),
    ("Sarah Lee",     "USB-C Cable",           7.99,  "Electronics"),
]

# 1) Calculate total revenue per product category and store it in a dictionary
category_revenue = {}  # dict: category -> total revenue

for (cust, product, price, category) in orders:
    if category not in category_revenue:
        category_revenue[category] = 0.0
    category_revenue[category] += price

# 2) Extract unique products from all orders using a set
unique_products = set()
for (_, product, _, _) in orders:
    unique_products.add(product)

# 3) Use a list comprehension to find all customers who purchased electronics
customers_electronics = list({
    cust for (cust, product, price, category) in orders
    if category == "Electronics"
})
# Using a set inside the comprehension ensures uniqueness; then we convert to list.

# 4) Identify the top three highest-spending customers using sorting
#    First, aggregate total spend per customer, then sort by spend (descending).
customer_totals = {}
for (cust, product, price, category) in orders:
    if cust not in customer_totals:
        customer_totals[cust] = 0.0
    customer_totals[cust] += price

# Sort by total spend, descending, and take top 3
top_three_customers = sorted(
    customer_totals.items(),
    key=lambda kv: kv[1],
    reverse=True
)[:3]

# ---------- Display results ----------
print("Total revenue per category:")
for cat, rev in category_revenue.items():
    print(f"  - {cat}: ${rev:,.2f}")

print("\nUnique products:")
for p in sorted(unique_products):
    print(f"  - {p}")

print("\nCustomers who purchased electronics:")
for cust in customers_electronics:
    print(f"  - {cust}")

print("\nTop three highest-spending customers:")
for cust, total in top_three_customers:
    print(f"  - {cust}: ${total:,.2f}")


Total revenue per category:
  - Electronics: $103.96
  - Clothing: $138.00
  - Home Essentials: $93.98

Unique products:
  - Bluetooth Headphones
  - Coffee Maker
  - Hoodie
  - Pillow Set (2)
  - Sneakers
  - USB-C Cable

Customers who purchased electronics:
  - Alex Johnson
  - Sarah Lee
  - Michael Brown

Top three highest-spending customers:
  - Maria Gomez: $99.00
  - Michael Brown: $77.49
  - David Chen: $59.00


**5. Organize and display data**

• Print a summary of each customer’s total spending and their classification

• Use set operations to find customers who purchased from multiple categories

• Identify common customers who bought both electronics and clothing

In [ ]:

# --- Sample orders (customer, product, price, category) ---
orders = [
    ("Alex Johnson",  "USB-C Cable",           9.99,  "Electronics"),
    ("Priya Singh",   "Hoodie",               39.00,  "Clothing"),
    ("David Chen",    "Coffee Maker",         59.00,  "Home Essentials"),
    ("Maria Gomez",   "Sneakers",             99.00,  "Clothing"),
    ("Sarah Lee",     "Pillow Set (2)",       17.99,  "Home Essentials"),
    ("Michael Brown", "Bluetooth Headphones", 69.00,  "Electronics"),
    ("Michael Brown", "USB-C Cable",           8.49,  "Electronics"),
    ("Alex Johnson",  "Pillow Set (2)",       16.99,  "Home Essentials"),
    ("Alex Johnson",  "USB-C Cable",           8.49,  "Electronics"),
    ("Sarah Lee",     "USB-C Cable",           7.99,  "Electronics"),
]

# --- 1) Compute total spending per customer ---
customer_totals = {}  # {customer -> total spend}
customer_categories = {}  # {customer -> set of categories seen}

for cust, product, price, category in orders:
    # totals
    if cust not in customer_totals:
        customer_totals[cust] = 0.0
    customer_totals[cust] += price

    # categories per customer
    if cust not in customer_categories:
        customer_categories[cust] = set()
    customer_categories[cust].add(category)

# --- 2) Classify customers by spend ---
# High-value: > $100
# Moderate:   $50–$100 (inclusive)
# Low-value:  < $50
def classify(total):
    if total > 100.0:
        return "High-value"
    elif 50.0 <= total <= 100.0:
        return "Moderate"
    return "Low-value"

customer_segment = {cust: classify(total) for cust, total in customer_totals.items()}

# --- 3) Print summary of each customer's total spending and classification ---
print("Customer Spending Summary")
print("-------------------------")
for cust in sorted(customer_totals.keys()):
    print(f"{cust:15s}  Total: ${customer_totals[cust]:6.2f}  →  {customer_segment[cust]}")

# --- 4) Use set operations to find customers who purchased from multiple categories ---
customers_multiple_categories = {cust for cust, cats in customer_categories.items() if len(cats) > 1}

print("\nCustomers who purchased from multiple categories:")
if customers_multiple_categories:
    for cust in sorted(customers_multiple_categories):
        cats = ", ".join(sorted(customer_categories[cust]))
        print(f"  - {cust} (categories: {cats})")
else:
    print("  (none)")

# --- 5) Identify customers who bought BOTH Electronics and Clothing (set intersection) ---
customers_electronics = {cust for cust, cats in customer_categories.items() if "Electronics" in cats}
customers_clothing    = {cust for cust, cats in customer_categories.items() if "Clothing" in cats}

customers_both_elec_and_cloth = customers_electronics & customers_clothing  # intersection

print("\nCustomers who bought BOTH Electronics and Clothing:")
if customers_both_elec_and_cloth:
    for cust in sorted(customers_both_elec_and_cloth):
        print(f"  - {cust}")
else:
    print("  (none)")


Customer Spending Summary
-------------------------
Alex Johnson     Total: $ 35.47  →  Low-value
David Chen       Total: $ 59.00  →  Moderate
Maria Gomez      Total: $ 99.00  →  Moderate
Michael Brown    Total: $ 77.49  →  Moderate
Priya Singh      Total: $ 39.00  →  Low-value
Sarah Lee        Total: $ 25.98  →  Low-value

Customers who purchased from multiple categories:
  - Alex Johnson (categories: Electronics, Home Essentials)
  - Sarah Lee (categories: Electronics, Home Essentials)

Customers who bought BOTH Electronics and Clothing:
  (none)


In [ ]:

"""
Customer Order Processing & Insights (built-ins only)

Data structures:
- customers: list[str]
- products: dict[str, dict]          # product_name -> {category, base_price}
- orders:   list[tuple]               # (customer_name, product_name, price_paid, category)

Features:
- Retrieve/modify using dict methods
- Spending classification loops
- Set operations for common/unique products across categories
- Insights: high-value customers, frequently purchased products, category-wise sales trends
"""

# -----------------------------
# 1) Store customer order data
# -----------------------------
customers = [
    "Alex Johnson",
    "Priya Singh",
    "David Chen",
    "Maria Gomez",
    "Sarah Lee",
    "Michael Brown",
    "Emily Davis",
]

# Master catalog: product -> (category, base_price)
products = {
    "USB-C Cable":           {"category": "Electronics",    "base_price": 9.99},
    "Bluetooth Headphones":  {"category": "Electronics",    "base_price": 79.00},
    "LED TV 55\"":           {"category": "Electronics",    "base_price": 579.00},

    "Hoodie":                {"category": "Clothing",       "base_price": 39.00},
    "Sneakers":              {"category": "Clothing",       "base_price": 99.00},
    "Jeans":                 {"category": "Clothing",       "base_price": 55.00},

    "Coffee Maker":          {"category": "Home Essentials","base_price": 69.00},
    "Pillow Set (2)":        {"category": "Home Essentials","base_price": 17.99},
    "Air Fryer":             {"category": "Home Essentials","base_price": 95.00},
}

# Orders as tuples (customer, product, price_paid, category)
# (Category is repeated here for quick filtering; alternatively derive from products dict)
orders = [
    ("Alex Johnson",  "USB-C Cable",           9.99,  "Electronics"),
    ("Alex Johnson",  "Pillow Set (2)",       16.99,  "Home Essentials"),
    ("Alex Johnson",  "USB-C Cable",           8.49,  "Electronics"),

    ("Priya Singh",   "Hoodie",               39.00,  "Clothing"),

    ("David Chen",    "Coffee Maker",         69.00,  "Home Essentials"),

    ("Maria Gomez",   "Sneakers",             99.00,  "Clothing"),

    ("Sarah Lee",     "Pillow Set (2)",       17.99,  "Home Essentials"),
    ("Sarah Lee",     "USB-C Cable",           7.99,  "Electronics"),

    ("Michael Brown", "Bluetooth Headphones", 79.00,  "Electronics"),
    ("Michael Brown", "USB-C Cable",           8.49,  "Electronics"),

    ("Emily Davis",   "Air Fryer",            95.00,  "Home Essentials"),
    ("Emily Davis",   "Jeans",                49.00,  "Clothing"),
]

# Build a dictionary: customer -> list of order tuples
customer_orders = {}
for cust, product, price, category in orders:
    customer_orders.setdefault(cust, []).append((product, price, category))


# ---------------------------------------------------
# 2) Retrieve and modify customer records (dict methods)
# ---------------------------------------------------
# Retrieve safely using get()
alex_orders = customer_orders.get("Alex Johnson", [])
# Add a new order using setdefault (ensures list exists)
customer_orders.setdefault("Alex Johnson", []).append(("LED TV 55\"", 579.00, "Electronics"))

# Update a customer record using update()
# (example: replace Emily's orders with a merged set—demo only)
merged_emily_orders = customer_orders.get("Emily Davis", []) + [("USB-C Cable", 9.49, "Electronics")]
customer_orders.update({"Emily Davis": merged_emily_orders})

# Remove a customer or a specific record using pop()
# (demo: remove Priya's last order if needed)
if customer_orders.get("Priya Singh"):
    customer_orders["Priya Singh"].pop()  # remove last order tuple

# Dictionary introspection
all_customers_in_dict = list(customer_orders.keys())
all_orders_values     = list(customer_orders.values())
all_items_view        = list(customer_orders.items())


# ---------------------------------------------------
# 3) Classification & analysis (using loops & sets)
# ---------------------------------------------------
# 3a) Compute total spending per customer
customer_totals = {}  # customer -> total_spend
customer_categories = {}  # customer -> set(categories purchased)

for cust, cust_orders in customer_orders.items():
    total = 0.0
    cats = set()
    for product, price, category in cust_orders:
        total += price
        cats.add(category)
    customer_totals[cust] = total
    customer_categories[cust] = cats

# 3b) Categorize customers by total spending
# Thresholds: High (>100), Moderate (50–100), Low (<50)
def classify_spend(total: float) -> str:
    if total > 100.0:
        return "High-value"


The final deliverable will be a detailed report summarizing customer classifications, total
sales per category, and key insights about purchase behavior. This project demonstrates
how Python’s data structures can be used to analyze real-world e-commerce data,
helping businesses make informed decisions.

**E-COMMERCE INSIGHTS REPORT**

---------------------------------------------

Total orders: 15

Total revenue: $4,769.31

Total profit: $2,091.31

Average order value (AOV): $317.95

Repeat purchase rate: 75.0%

**Top products by units sold:**
  - USB-C Cable (P103) — Units: 10, Revenue: $86.90, Profit: $56.90
  - Pillow Set (2) (P303) — Units: 8, Revenue: $143.92, Profit: $79.92
  - Socks (5-pack) (P203) — Units: 6, Revenue: $52.49, Profit: $28.49
  - Hoodie (P200) — Units: 4, Revenue: $146.00, Profit: $74.00
  - Coffee Maker (P302) — Units: 3, Revenue: $207.00, Profit: $141.00
  - Jeans (P201) — Units: 3, Revenue: $159.00, Profit: $93.00
  - Air Fryer (P300) — Units: 2, Revenue: $184.00, Profit: $114.00
  - Bluetooth Headphones (P101) — Units: 2, Revenue: $148.00, Profit: $78.00

**Most frequently purchased products (by number of orders):**
  - USB-C Cable (P103) — Appears in 5 orders, Units: 10
  - Hoodie (P200) — Appears in 3 orders, Units: 4
  - Socks (5-pack) (P203) — Appears in 3 orders, Units: 6
  - Coffee Maker (P302) — Appears in 3 orders, Units: 3
  - Pillow Set (2) (P303) — Appears in 3 orders, Units: 8
  - Air Fryer (P300) — Appears in 2 orders, Units: 2
  - Bluetooth Headphones (P101) — Appears in 2 orders, Units: 2
  - Sneakers (P202) — Appears in 2 orders, Units: 2

**Customer spending tiers:**
  - Bronze: 2 customers (range $0.00–$250.00)
  - Silver: 3 customers (range $250.00–$750.00)
  - Gold: 3 customers (range $750.00–$1,500.00)
  - Platinum: 0 customers (range $1,500.00–∞)

  **Top customers by total spend:**

    · Michael Brown (C006) — $1,383.98 in 2 orders — Tier: Gold
    · Alex Johnson (C001) — $811.41 in 3 orders — Tier: Gold
    · Maria Gomez (C004) — $753.98 in 2 orders — Tier: Gold
    · Priya Singh (C002) — $666.49 in 2 orders — Tier: Silver
    · David Chen (C003) — $455.98 in 2 orders — Tier: Silver

**Most profitable categories:**
  - Electronics: Profit $1,180.90 on Revenue $3,330.90 — Units: 16, Margin: 35.5%
  - Home Essentials: Profit $582.92 on Revenue $892.92 — Units: 15, Margin: 65.3%
  - Clothing: Profit $327.49 on Revenue $545.49 — Units: 15, Margin: 60.0%

**Inventory suggestions (heuristic):**
  - USB-C Cable (P103): Units 10, Margin 65.5%, In 5 orders
  - Pillow Set (2) (P303): Units 8, Margin 55.5%, In 3 orders
  - Socks (5-pack) (P203): Units 6, Margin 54.3%, In 3 orders
  - Hoodie (P200): Units 4, Margin 50.7%, In 3 orders